### PR TITLE
Fix #1225

### LINKED ISSUE #1225: Too many entries in history for sub-namespaces
E.g.

```unison
myLibrary.f x = x + 9
myLibrary.g x = x + 10
myLibrary.h x = x + 3
```

```ucm
.> add
.> history
.> cd myLibrary
.myLibrary> history
```

The first `history` command (from the root) has one entry for all three functions, but the second command (issued from the `myLibrary` namespace) shows three entries; one for each function.

### PR DESCRIPTION
**Pull request by pchiusano**
This is all set. To be merged after #1337.

## Overview

Fixes #1225, where a separate history entry was being created for each definition being added to a namespace. Now uses a post-order traversal of the branch as suggested by @aryairani, rebuilding from the bottom up and visiting each node exactly once. This avoids creating excessive history nodes. See the changed output for `link.md` transcript.

## Implementation notes

Change was very surgical, to `Branch.stepManyAt0` and `Branch.stepManyAt0M`.

## Interesting/controversial decisions

None that I can think of. There was an interesting change to the `delete.output.md`, but I stared at it a while and convinced myself the old output was wrong (annotated below). I don't totally understand why this change has fixed that bug though.

Also important note: this affects hashes for other transcripts just due to there being fewer steps in the history. All the transcripts still behave correctly though.

## Test coverage

Existing transcripts provide pretty good test coverage I think. I'd implemented it incorrectly at first and got lots of failures. :) 

I'll see if I can think of some more specific tests though.

### CODE REVIEW

File: unison-src/transcripts/link.output.md [hunk 1]
```diff
@@ -196,27 +196,15 @@ myLibrary.h x = x + 3
...
-    + Adds / updates:
-    
-      g
-  
-  ⊙ #seb6uuoocl
-  
-    + Adds / updates:
-    
-      f
+      f g h
```
Comment (pchiusano) [inline unison-src/transcripts/link.output.md]: Here's where we see the new behavior: `f`, `g`, and `h` are all added in one go, rather than 3 separate steps. This then impacts later namespace hashes.

File: unison-src/transcripts/reflog.output.md [hunk 2]
```diff
@@ -65,16 +65,16 @@ y = 2
...
+    `reset-root #d7reeo8lc4`  to reset the root namespace and
                               its history to that of the
                               specified namespace.
   
-  1. #bqd4d6mrnk : add
-  2. #onc5fpvvb7 : add
-  3. #upd8bubg9e : builtins.merge
+  1. #up853v4413 : add
+  2. #d7reeo8lc4 : add
+  3. #i5f8gcfb14 : builtins.merge
```
Comment (pchiusano) [inline unison-src/transcripts/reflog.output.md]: I'm not seeing an obvious reason why this hash should be different so going to investigate. `builtins.merge` doesn't use `stepManyAt` that I can see.

File: unison-src/transcripts/delete.output.md [hunk 3]
```diff
@@ -182,10 +188,6 @@ type Foo = Foo Boolean
     5. ┌ Foo.Foo#d97e0jhkmd#0 : Nat -> Foo
     6. └ Foo.Foo#gq9inhvg9h#0 : Boolean -> b.Foo
   
-  Added definitions:
-  
-    7. foo : Nat
-
```
Comment (pchiusano) [inline unison-src/transcripts/delete.output.md]: This old output was incorrect: `foo` should have been deleted in the `a` namespace, so it shouldn't be resurfacing here.

File: unison-src/transcripts/reflog.output.md (hunk 2 above)
Comment (pchiusano) [reply unison-src/transcripts/reflog.output.md]: Okay, it does actually call `stepManyAt` (indirectly via `BranchUtils.addFromNames0`). So mystery solved.

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 4]
```diff
@@ -719,32 +720,60 @@ modifyAtM path f b = case Path.uncons path of
...
-
-stepManyAt0M :: (Monad m, Monad n, Foldable f)
+stepManyAt0 actions b = go (first Path.toList <$> toList actions) b where
+  go :: [([NameSegment], Branch0 m -> Branch0 m)] -> Branch0 m -> Branch0 m
+  go actions b = let 
+    -- combines the functions that apply to this level of the tree
+    currentAction b = foldl' (\b f -> f b) b [ f | ([], f) <- actions ]
+
+    -- groups the actions based on the child they apply to
+    childActions :: Map NameSegment [([NameSegment], Branch0 m -> Branch0 m)]
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: What's the `[NameSegment]`?

File: unison-src/transcripts/delete.output.md (hunk 3 above)
Comment (aryairani) [reply unison-src/transcripts/delete.output.md]: Seems like a merge error.  We had `.a.foo = 1` and `.b.foo = 2`, then we merged `.b` into `.a`, and deleted `.a.foo`.  So we have no `.a.foo` but we still have `.b.foo`.  Then we merged `.b` into `.a` again, and had previously ended up  with `.a.foo = 2` (presumably) again.

But why did it change now?

Since the version of `.b` that contained `foo` was already part of the history of `.a`, it seems like it shouldn't have been added again.  I wouldn't have been surprised if this changed with some `lca` or `merge` thing, but I still can't really imagine the issue and don't get what it has to do with this PR.

Comment (pchiusano) [reply unison-src/transcripts/delete.output.md]: I think I get it: old implementation would have a chain of history for each definition added, and moreover the definitions in the chain would be in an arbitrary order (or maybe alphabetical). This then affects LCA calculation and the merge result. I’ll see if I can confirm this hypothesis.

File: parser-typechecker/src/Unison/Codebase/Branch.hs (hunk 4 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: It’s just the Path - notice the initial call to go which converts the paths to lists of NameSegment.

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Oh, I see.  What do you think about leaving it as `Path` for self-documentation purposes?  Maybe the cons (`<|`) and snoc (`|>`) patterns will help?

File: unison-src/transcripts/delete.output.md (hunk 3 above)
Comment (aryairani) [reply unison-src/transcripts/delete.output.md]: Interesting, okay.

Comment (pchiusano) [reply unison-src/transcripts/delete.output.md]: Okay, got to the bottom of this and it's not what I thought. The difference in behavior is because the new code leaves around an empty sub-branch which remembers the causal history that led to that sub-branch being empty. The old code was over-eager about deleting sub-branches that became empty. Here's [transcript output, showing the history](https://gist.github.com/pchiusano/9edb8e546149bce49ef1574cdf195577)

The key is that in the `a` namespace in this PR, history after the delete looks like what you'd expect:

```
.a> history

  Note: The most recent namespace hash is immediately below this
        message.
  
  ⊙ #gcq0ml6diu
  
    - Deletes:
    
      foo#0ja foo#jk1
  
  This segment of history starts with a merge. Use
  `history #som3n4m3space` to view history starting from a given
  namespace hash.
  
  #biv51qeect
  #u8lsnrulgl
  ⑂
  ⊙ #l24km9hmve
```

This reflects that a merge occurred and then `foo` (both versions) were then deleted. Whereas in `master`, this same portion of the transcript looks like:

```
.a> history

  ☝️  The namespace .a is empty.
```

As a result, when doing the subsequent merge of `b` into `a`, `master` has "forgotten" that at an earlier point in `a` history, `foo` was deleted. So it treats the `foo` as a new definition and adds it.

So this branch has the correct behavior, however I was curious what made the old code so eager to delete empty branches. Old code was defined in terms of:

https://github.com/unisonweb/unison/blob/b5cae909fc15639d8f6e86fc86b83141700a2675/parser-typechecker/src/Unison/Codebase/Branch.hs#L689-L696

Which as you can see just deletes the key if you ever go to insert something whose `head` is `Branch.empty0`.

So basically, this PR just happens to work due better in this instance due to a coincidence of the implementation. I'd like to understand better when it is safe to delete empty sub-branches. This seems to be a case where being overeager to delete them leads to worse merge results. I suspect you should just never delete empty sub-branches, unless the user has literally typed `delete.namespace foo.bar`. You shouldn't just delete all the history just because the namespace is currently empty... that would be a little like deleting a directory automatically once you delete the last file.

Comment (aryairani) [reply unison-src/transcripts/delete.output.md]: Doh, good work.  Not sold on the relevance of the directory analogy, but agree we shouldn't be deleting info needed to merge correctly :)

File: parser-typechecker/src/Unison/Codebase/Path.hs [hunk 5]
```diff
@@ -312,6 +312,9 @@ relativeToAncestor (Path a) (Path b) = case (a, b) of
   _ -> (empty, Path a, Path b)
 
 pattern Parent h t = Path (NameSegment h :<| t)
+pattern Cons h t <- Path (h :<| (Path -> t)) where
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Path.hs]: Instructions unclear, sorry — I suggest using the cons pattern `(<|)` that comes with Control.Lens, for which we have instances below; though if you think it’s better to have an explicit one, that’s fine by me too.

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Path.hs]: Struggling to understand what `Parent` is there too, relative to Cons and Snoc, though I acknowledge it’s not part of this PR.

Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Path.hs]: Yeah, not sure where `Parent` is used, seems a bit weird to me. I'll try out the `<|` pattern, still want the `Empty` pattern though.

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Path.hs]: Just wanted to point out that those instances existed, but I'm good with your choice at that point.

**Comment by aryairani** [comment]
and switch back to `updateChildren` if you want.

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 6]
```diff
@@ -725,52 +725,52 @@ stepManyAt0 :: forall f m . (Applicative m, Foldable f)
...
+            step (go actions) (Map.findWithDefault empty seg m0)
+          Nothing -> Map.findWithDefault empty seg m0
+    in currentAction $ over children stepChildren b
 
 stepManyAt0M :: forall m n f . (Monad m, Monad n, Foldable f)
              => f (Path, Branch0 m -> n (Branch0 m))
              -> Branch0 m -> n (Branch0 m)
 stepManyAt0M actions b = go (toList actions) b where
   go :: [(Path, Branch0 m -> n (Branch0 m))] -> Branch0 m -> n (Branch0 m)
-  go actions b = let
```
Comment (pchiusano) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: kill trailing whitespace

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 7]
```diff
@@ -725,52 +725,52 @@ stepManyAt0 :: forall f m . (Applicative m, Foldable f)
...
              => f (Path, Branch0 m -> n (Branch0 m))
              -> Branch0 m -> n (Branch0 m)
 stepManyAt0M actions b = go (toList actions) b where
   go :: [(Path, Branch0 m -> n (Branch0 m))] -> Branch0 m -> n (Branch0 m)
-  go actions b = let 
+  go actions b = let
     currentAction b = foldM (\b f -> f b) b [ f | (Path.Empty, f) <- actions ]
 
     childActions :: Map NameSegment [(Path, Branch0 m -> n (Branch0 m))]
-    childActions =
```
Comment (pchiusano) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: kill trailing whitespace

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 8]
```diff
@@ -719,32 +719,62 @@ modifyAtM path f b = case Path.uncons path of
...
+    currentAction b = foldl' (\b f -> f b) b [ f | (Path.Empty, f) <- actions ]
+
+    -- groups the actions based on the child they apply to
+    childActions :: Map NameSegment [(Path, Branch0 m -> Branch0 m)]
+    childActions =
+      List.multimap [ (seg, (rest,f)) | (seg :< rest, f) <- actions ]
+
+    -- alters the children of `b` based on the `childActions` map
+    stepChildren :: Map NameSegment (Branch m) -> Map NameSegment (Branch m)
+    stepChildren m0 = foldl' g Map.empty (Map.keysSet m0 <> Map.keysSet childActions)
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: I might rename `m0` to `children`, since that's whats passed in.

Then... we're folding over the names of the existing children + the names that have actions associated (may correspond to existing children or to children that don't yet exist).

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: `actions` gets partitioned between `currentAction` (and preemptively composed together), and `childActions`.

Why do we fold over `keysSet m0`?

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 9]
```diff
@@ -719,32 +719,62 @@ modifyAtM path f b = case Path.uncons path of
...
+      -- this skips adding any empty children, which could happen depending
+      -- on the action functions provided (think: `b -> Branch.empty0`)
+      insert seg child m = if isEmpty child then m else Map.insert seg child m
+      g m seg = insert seg child m where
+        child = case Map.lookup seg childActions of
+          Just actions ->
+            -- this `findWithDefault` is important, allows the stepManyAt
+            -- to create new children at deeper paths than exist in the
+            -- original Branch
+            step (go actions) (Map.findWithDefault empty seg m0)
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: `(go actions)` aggregates all of the actions at this level and below into a single action. :+1:

File: parser-typechecker/src/Unison/Codebase/Branch.hs (hunk 8 above)
Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Notice the `foldl'` starts with `Map.empty`.

If we just folded over `keysSet childActions`, the `foldl'` would delete all sub-branches that the `childActions` didn't touch... likewise, if we just folded over `keysSet m0`, then we wouldn't be able to have actions that added new child namespaces (I made this mistake in an earlier implementation and lots of tests failed).

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 10]
```diff
@@ -719,32 +719,62 @@ modifyAtM path f b = case Path.uncons path of
...
+      where
+      insert seg child m = if isEmpty child then m else Map.insert seg child m
+      g m seg = do
+        child <- case Map.lookup seg childActions of
+          Just actions -> stepM (go actions) (Map.findWithDefault empty seg m0)
+          Nothing -> pure $ Map.findWithDefault empty seg m0
+        pure $ insert seg child m
+    in do
+      c2 <- stepChildren (view children b)
+      currentAction (set children c2 b)
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: Is there a tidy, correct way to share an implementation between `stepManyAt0` and `stepManyAt0M`

Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Yeah, I think it'll work to define both in terms of the monadic one, I'll try that out in separate PR.

File: parser-typechecker/src/Unison/Codebase/Branch.hs (hunk 8 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Oh, hm... <stares more>

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: What about `fold' g m0 (Map.keysSet childActions)`? instead of starting with `Map.empty`?

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: So like there's Map.lookup on the childActions, below.  If it's `Just`, then we call step.  If it's `Nothing`, then that means it's a `seg` that's in `m0` aka `existingChildren`, but not in `childActions`.

When it's `Nothing`, we `Map.findWithDefault empty seg m0` which — it will never use the default, because `seg` is definitely in `m0`, as we know it isn't in `childActions`, and it had to have come from somewhere.

So I think maybe that fold in my previous message will do the trick; leaving such children unchanged from m0, which we're doing anyway, "with extra steps", as Rick and Morty would say.

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Moreover, we could get rid of that `Map.lookup` I think, since everything we need is in `childActions`; correct me if I'm wrong.

Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Yeah these are good improvements... agree with all of them and PR coming shortly. Maybe next time do this as part of the initial review though. :)

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Yeah, my bad. 😓 

I think a sort of mental or definitional shift was needed on my end to decide/realize what code review should be.  I think it's this. 😅

Comment (pchiusano) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: Np :) It is more work to do a detailed review of implementation details and suggest improvements... so there's some cost-benefit there and I wouldn't say it always needs to be done. Not sure what general guidance is on when to do it.
